### PR TITLE
[HDRP] Added a RenderGraph pass that resets the camera size after the dynamic res upscale

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
Fix https://fogbugz.unity3d.com/f/cases/1299194/ and https://forum.unity.com/threads/viewport-is-incorrect-in-a-custom-pass-running-in-afterpostprocess-when-using-dynamic-resolution.1008029/

Context: When the dynamic resolution is enabled, HDRP renders everything in a smaller viewport up to the final post-process pass which is just after the "After PP" post-process injection point.  
After post-process, opaque materials are rendered before post processes in an offscreen target and composited back during this final pass.  
So the custom pass after post-process is the only pass that renders something after the final pass that does the upscaling of the image for dynamic res.

Because of that, the after post-process custom passes are rendered with a smaller viewport than expected and there is no upscale pass after this one so it leads scale issue during the custom pass rendering.

To fix this, I added a pass that resets the camera size after the dynamic res upscale pass, so everything rendered after that is using the correct scale.

---
### Testing status
Tested that there is no more scale issue with a simple after post-process custom pass and dynamic res enabled (with a fixed resolution scale multiplier).  
Also verified that there is no CG alloc in the RG pass.

https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/hd%252Ffix%252Fafter-post-process-dynamic-res/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/4886416/job

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs [hunk 1]
```diff
@@ -1535,6 +1537,33 @@ void RenderGizmos(RenderGraph renderGraph, HDCamera hdCamera, TextureHandle colo
             return executed;
         }
 
+        class BindResetCameraSizeForAfterPostProcess
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:1540]: Most if not all other pass data are called XXXPassData so I think we should rename this for consistency.

**Review by JulienIgnace-Unity — CHANGES_REQUESTED** [review]
Requested a rename but otherwise lgtm. Hopefully we won't have unforseen side effects. Tagging @FrancescoC-unity in case you can think of anything like that.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs (hunk 1 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs]: fixed

**Review by JulienIgnace-Unity — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs [hunk 2]
```diff
@@ -1535,6 +1537,33 @@ void RenderGizmos(RenderGraph renderGraph, HDCamera hdCamera, TextureHandle colo
...
 
+        class ResetCameraSizeForAfterPostProcessPassData
+        {
+            public HDCamera hdCamera;
+            public ShaderVariablesGlobal shaderVariablesGlobal;
+        }
+
+        void ResetCameraSizeForAfterPostProcess(RenderGraph renderGraph, HDCamera hdCamera, CommandBuffer commandBuffer)
+        {
+            if (hdCamera.allowDynamicResolution && HDUtils.hdrpSettings.dynamicResolutionSettings.enabled)
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:1548]: Better to go through the Dynamic resolution handler method: DynamicResolutionEnabled ( `DynamicResolutionHandler.instance.DynamicResolutionEnabled ` )

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs [hunk 3]
```diff
@@ -1535,6 +1537,33 @@ void RenderGizmos(RenderGraph renderGraph, HDCamera hdCamera, TextureHandle colo
...
+                using (var builder = renderGraph.AddRenderPass("Reset Camera Size After Post Process", out ResetCameraSizeForAfterPostProcessPassData passData))
+                {
+                    passData.hdCamera = hdCamera;
+                    passData.shaderVariablesGlobal = m_ShaderVariablesGlobalCB;
+                    builder.AllowPassCulling(false);
+
+                    builder.SetRenderFunc(
+                        (ResetCameraSizeForAfterPostProcessPassData data, RenderGraphContext ctx) =>
+                        {
+                            data.shaderVariablesGlobal._ScreenSize = new Vector4(data.hdCamera.finalViewport.width, data.hdCamera.finalViewport.height, 1.0f / data.hdCamera.finalViewport.width, 1.0f / data.hdCamera.finalViewport.height);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:1559]: Do we need updating the RTHandle scale stuff before pushing the cbuffer?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs (hunk 2 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:1548]: Thanks, fixed :)

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs (hunk 3 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:1566]: Good catch! indeed the RTHandleScale needs to be updated if we want to sample low-res buffers from before the upscale

**Review by sebastienlagarde — APPROVED** [review]

**Review by FrancescoC-unity — APPROVED** [review]

**Pull request merged**